// @effect/vitest/dist/esm/internal.js
import * as Arbitrary from "effect/Arbitrary";
import * as Cause from "effect/Cause";
import * as Duration from "effect/Duration";
import * as Effect from "effect/Effect";
import * as Equal from "effect/Equal";
import * as Exit from "effect/Exit";
import * as Fiber from "effect/Fiber";
import { flow, identity, pipe } from "effect/Function";
import * as Layer from "effect/Layer";
import * as Logger from "effect/Logger";
import * as Schedule from "effect/Schedule";
import * as Schema from "effect/Schema";
import * as Scope from "effect/Scope";
import * as TestEnvironment from "effect/TestContext";
import * as Utils from "effect/Utils";
import fc from "fast-check";
import * as V from "vitest";
const runPromise = ctx => effect => Effect.gen(function* () {
  const exitFiber = yield* Effect.fork(Effect.exit(effect));
  ctx?.onTestFinished(() => Fiber.interrupt(exitFiber).pipe(Effect.asVoid, Effect.runPromise));
  const exit = yield* Fiber.join(exitFiber);
  if (Exit.isSuccess(exit)) {
    return () => exit.value;
  } else {
    const errors = Cause.prettyErrors(exit.cause);
    for (let i = 1; i < errors.length; i++) {
      yield* Effect.logError(errors[i]);
    }
    return () => {
      throw errors[0];
    };
  }
}).pipe(Effect.runPromise).then(f => f());
/** @internal */
const runTest = ctx => effect => runPromise(ctx)(effect);
/** @internal */
const TestEnv = /*#__PURE__*/TestEnvironment.TestContext.pipe( /*#__PURE__*/Layer.provide( /*#__PURE__*/Logger.remove(Logger.defaultLogger)));
/** @internal */
function customTester(a, b, customTesters) {
  if (!Equal.isEqual(a) || !Equal.isEqual(b)) {
    return undefined;
  }
  return Utils.structuralRegion(() => Equal.equals(a, b), (x, y) => this.equals(x, y, customTesters.filter(t => t !== customTester)));
}
/** @internal */
export const addEqualityTesters = () => {
  V.expect.addEqualityTesters([customTester]);
};
/** @internal */
const makeTester = mapEffect => {
  const run = (ctx, args, self) => pipe(Effect.suspend(() => self(...args)), mapEffect, runTest(ctx));
  const f = (name, self, timeout) => V.it(name, ctx => run(ctx, [ctx], self), timeout);
  const skip = (name, self, timeout) => V.it.skip(name, ctx => run(ctx, [ctx], self), timeout);
  const skipIf = condition => (name, self, timeout) => V.it.skipIf(condition)(name, ctx => run(ctx, [ctx], self), timeout);
  const runIf = condition => (name, self, timeout) => V.it.runIf(condition)(name, ctx => run(ctx, [ctx], self), timeout);
  const only = (name, self, timeout) => V.it.only(name, ctx => run(ctx, [ctx], self), timeout);
  const each = cases => (name, self, timeout) => V.it.for(cases)(name, typeof timeout === "number" ? {
    timeout
  } : timeout ?? {}, (args, ctx) => run(ctx, [args], self));
  const prop = (name, arbitraries, self, timeout) => {
    if (Array.isArray(arbitraries)) {
      const arbs = arbitraries.map(arbitrary => Schema.isSchema(arbitrary) ? Arbitrary.make(arbitrary) : arbitrary);
      return V.it(name,
      // @ts-ignore
      ctx => fc.assert(fc.asyncProperty(...arbs, (...as) => run(ctx, [as, ctx], self))), timeout);
    }
    const arbs = fc.record(Object.keys(arbitraries).reduce(function (result, key) {
      result[key] = Schema.isSchema(arbitraries[key]) ? Arbitrary.make(arbitraries[key]) : arbitraries[key];
      return result;
    }, {}));
    return V.it(name,
    // @ts-ignore
    ctx => fc.assert(fc.asyncProperty(arbs, (...as) => run(ctx, [as[0], ctx], self))), timeout);
  };
  return Object.assign(f, {
    skip,
    skipIf,
    runIf,
    only,
    each,
    prop
  });
};
export const prop = (name, arbitraries, self, timeout) => {
  if (Array.isArray(arbitraries)) {
    const arbs = arbitraries.map(arbitrary => Schema.isSchema(arbitrary) ? Arbitrary.make(arbitrary) : arbitrary);
    return V.it(name,
    // @ts-ignore
    ctx => fc.assert(fc.property(...arbs, (...as) => self(as, ctx))), timeout);
  }
  const arbs = fc.record(Object.keys(arbitraries).reduce(function (result, key) {
    result[key] = Schema.isSchema(arbitraries[key]) ? Arbitrary.make(arbitraries[key]) : arbitraries[key];
    return result;
  }, {}));
  return V.it(name,
  // @ts-ignore
  ctx => fc.assert(fc.property(arbs, (...as) => self(as[0], ctx))), timeout);
};
/** @internal */
export const layer = (layer_, options) => (...args) => {
  const memoMap = options?.memoMap ?? Effect.runSync(Layer.makeMemoMap);
  const scope = Effect.runSync(Scope.make());
  const runtimeEffect = Layer.toRuntimeWithMemoMap(layer_, memoMap).pipe(Scope.extend(scope), Effect.orDie, Effect.cached, Effect.runSync);
  const it = Object.assign(V.it, {
    effect: makeTester(effect => Effect.flatMap(runtimeEffect, runtime => effect.pipe(Effect.provide(runtime), Effect.provide(TestEnv)))),
    prop,
    scoped: makeTester(effect => Effect.flatMap(runtimeEffect, runtime => effect.pipe(Effect.scoped, Effect.provide(runtime), Effect.provide(TestEnv)))),
    live: makeTester(effect => Effect.flatMap(runtimeEffect, runtime => Effect.provide(effect, runtime))),
    scopedLive: makeTester(effect => Effect.flatMap(runtimeEffect, runtime => effect.pipe(Effect.scoped, Effect.provide(runtime)))),
    flakyTest,
    layer(nestedLayer, options) {
      return layer(Layer.provideMerge(nestedLayer, layer_), {
        ...options,
        memoMap
      });
    }
  });
  if (args.length === 1) {
    V.beforeAll(() => runPromise()(Effect.asVoid(runtimeEffect)), options?.timeout ? Duration.toMillis(options.timeout) : undefined);
    V.afterAll(() => runPromise()(Scope.close(scope, Exit.void)), options?.timeout ? Duration.toMillis(options.timeout) : undefined);
    return args[0](it);
  }
  return V.describe(args[0], () => {
    V.beforeAll(() => runPromise()(Effect.asVoid(runtimeEffect)), options?.timeout ? Duration.toMillis(options.timeout) : undefined);
    V.afterAll(() => runPromise()(Scope.close(scope, Exit.void)), options?.timeout ? Duration.toMillis(options.timeout) : undefined);
    return args[1](it);
  });
};
/** @internal */
export const effect = /*#__PURE__*/makeTester( /*#__PURE__*/Effect.provide(TestEnv));
/** @internal */
export const scoped = /*#__PURE__*/makeTester( /*#__PURE__*/flow(Effect.scoped, /*#__PURE__*/Effect.provide(TestEnv)));
/** @internal */
export const live = /*#__PURE__*/makeTester(identity);
/** @internal */
export const scopedLive = /*#__PURE__*/makeTester(Effect.scoped);
/** @internal */
export const flakyTest = (self, timeout = Duration.seconds(30)) => pipe(Effect.catchAllDefect(self, Effect.fail), Effect.retry(pipe(Schedule.recurs(10), Schedule.compose(Schedule.elapsed), Schedule.whileOutput(Duration.lessThanOrEqualTo(timeout)))), Effect.orDie);
