// @effect/vitest/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  addEqualityTesters: true,
  effect: true,
  scoped: true,
  live: true,
  scopedLive: true,
  layer: true,
  flakyTest: true,
  prop: true,
  it: true
};
exports.scopedLive = exports.scoped = exports.prop = exports.live = exports.layer = exports.it = exports.flakyTest = exports.effect = exports.addEqualityTesters = void 0;
var V = _interopRequireWildcard(require("vitest"));
Object.keys(V).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === V[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return V[key];
    }
  });
});
var internal = _interopRequireWildcard(require("./internal.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 1.0.0
 */
const addEqualityTesters = exports.addEqualityTesters = internal.addEqualityTesters;
/**
 * @since 1.0.0
 */
const effect = exports.effect = internal.effect;
/**
 * @since 1.0.0
 */
const scoped = exports.scoped = internal.scoped;
/**
 * @since 1.0.0
 */
const live = exports.live = internal.live;
/**
 * @since 1.0.0
 */
const scopedLive = exports.scopedLive = internal.scopedLive;
/**
 * Share a `Layer` between multiple tests, optionally wrapping
 * the tests in a `describe` block if a name is provided.
 *
 * @since 1.0.0
 *
 * ```ts
 * import { expect, layer } from "@effect/vitest"
 * import { Context, Effect, Layer } from "effect"
 *
 * class Foo extends Context.Tag("Foo")<Foo, "foo">() {
 *   static Live = Layer.succeed(Foo, "foo")
 * }
 *
 * class Bar extends Context.Tag("Bar")<Bar, "bar">() {
 *   static Live = Layer.effect(
 *     Bar,
 *     Effect.map(Foo, () => "bar" as const)
 *   )
 * }
 *
 * layer(Foo.Live)("layer", (it) => {
 *   it.effect("adds context", () =>
 *     Effect.gen(function* () {
 *       const foo = yield* Foo
 *       expect(foo).toEqual("foo")
 *     })
 *   )
 *
 *   it.layer(Bar.Live)("nested", (it) => {
 *     it.effect("adds context", () =>
 *       Effect.gen(function* () {
 *         const foo = yield* Foo
 *         const bar = yield* Bar
 *         expect(foo).toEqual("foo")
 *         expect(bar).toEqual("bar")
 *       })
 *     )
 *   })
 * })
 * ```
 */
const layer = exports.layer = internal.layer;
/**
 * @since 1.0.0
 */
const flakyTest = exports.flakyTest = internal.flakyTest;
/**
 * @since 1.0.0
 */
const prop = exports.prop = internal.prop;
/**
 * @since 1.0.0
 */
/** @ignored */
const methods = {
  effect,
  live,
  flakyTest,
  scoped,
  scopedLive,
  layer,
  prop
};
/**
 * @since 1.0.0
 */
const it = exports.it = /*#__PURE__*/Object.assign(V.it, methods);
/**
 * @since 1.0.0
 */
