// @effect/vitest/dist/cjs/internal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.scopedLive = exports.scoped = exports.prop = exports.live = exports.layer = exports.flakyTest = exports.effect = exports.addEqualityTesters = void 0;
var Arbitrary = _interopRequireWildcard(require("effect/Arbitrary"));
var Cause = _interopRequireWildcard(require("effect/Cause"));
var Duration = _interopRequireWildcard(require("effect/Duration"));
var Effect = _interopRequireWildcard(require("effect/Effect"));
var Equal = _interopRequireWildcard(require("effect/Equal"));
var Exit = _interopRequireWildcard(require("effect/Exit"));
var Fiber = _interopRequireWildcard(require("effect/Fiber"));
var _Function = require("effect/Function");
var Layer = _interopRequireWildcard(require("effect/Layer"));
var Logger = _interopRequireWildcard(require("effect/Logger"));
var Schedule = _interopRequireWildcard(require("effect/Schedule"));
var Schema = _interopRequireWildcard(require("effect/Schema"));
var Scope = _interopRequireWildcard(require("effect/Scope"));
var TestEnvironment = _interopRequireWildcard(require("effect/TestContext"));
var Utils = _interopRequireWildcard(require("effect/Utils"));
var _fastCheck = _interopRequireDefault(require("fast-check"));
var V = _interopRequireWildcard(require("vitest"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const runPromise = ctx => effect => Effect.gen(function* () {
  const exitFiber = yield* Effect.fork(Effect.exit(effect));
  ctx?.onTestFinished(() => Fiber.interrupt(exitFiber).pipe(Effect.asVoid, Effect.runPromise));
  const exit = yield* Fiber.join(exitFiber);
  if (Exit.isSuccess(exit)) {
    return () => exit.value;
  } else {
    const errors = Cause.prettyErrors(exit.cause);
    for (let i = 1; i < errors.length; i++) {
      yield* Effect.logError(errors[i]);
    }
    return () => {
      throw errors[0];
    };
  }
}).pipe(Effect.runPromise).then(f => f());
/** @internal */
const runTest = ctx => effect => runPromise(ctx)(effect);
/** @internal */
const TestEnv = /*#__PURE__*/TestEnvironment.TestContext.pipe( /*#__PURE__*/Layer.provide( /*#__PURE__*/Logger.remove(Logger.defaultLogger)));
/** @internal */
function customTester(a, b, customTesters) {
  if (!Equal.isEqual(a) || !Equal.isEqual(b)) {
    return undefined;
  }
  return Utils.structuralRegion(() => Equal.equals(a, b), (x, y) => this.equals(x, y, customTesters.filter(t => t !== customTester)));
}
/** @internal */
const addEqualityTesters = () => {
  V.expect.addEqualityTesters([customTester]);
};
/** @internal */
exports.addEqualityTesters = addEqualityTesters;
const makeTester = mapEffect => {
  const run = (ctx, args, self) => (0, _Function.pipe)(Effect.suspend(() => self(...args)), mapEffect, runTest(ctx));
  const f = (name, self, timeout) => V.it(name, ctx => run(ctx, [ctx], self), timeout);
  const skip = (name, self, timeout) => V.it.skip(name, ctx => run(ctx, [ctx], self), timeout);
  const skipIf = condition => (name, self, timeout) => V.it.skipIf(condition)(name, ctx => run(ctx, [ctx], self), timeout);
  const runIf = condition => (name, self, timeout) => V.it.runIf(condition)(name, ctx => run(ctx, [ctx], self), timeout);
  const only = (name, self, timeout) => V.it.only(name, ctx => run(ctx, [ctx], self), timeout);
  const each = cases => (name, self, timeout) => V.it.for(cases)(name, typeof timeout === "number" ? {
    timeout
  } : timeout ?? {}, (args, ctx) => run(ctx, [args], self));
  const prop = (name, arbitraries, self, timeout) => {
    if (Array.isArray(arbitraries)) {
      const arbs = arbitraries.map(arbitrary => Schema.isSchema(arbitrary) ? Arbitrary.make(arbitrary) : arbitrary);
      return V.it(name,
      // @ts-ignore
      ctx => _fastCheck.default.assert(_fastCheck.default.asyncProperty(...arbs, (...as) => run(ctx, [as, ctx], self))), timeout);
    }
    const arbs = _fastCheck.default.record(Object.keys(arbitraries).reduce(function (result, key) {
      result[key] = Schema.isSchema(arbitraries[key]) ? Arbitrary.make(arbitraries[key]) : arbitraries[key];
      return result;
    }, {}));
    return V.it(name,
    // @ts-ignore
    ctx => _fastCheck.default.assert(_fastCheck.default.asyncProperty(arbs, (...as) => run(ctx, [as[0], ctx], self))), timeout);
  };
  return Object.assign(f, {
    skip,
    skipIf,
    runIf,
    only,
    each,
    prop
  });
};
const prop = (name, arbitraries, self, timeout) => {
  if (Array.isArray(arbitraries)) {
    const arbs = arbitraries.map(arbitrary => Schema.isSchema(arbitrary) ? Arbitrary.make(arbitrary) : arbitrary);
    return V.it(name,
    // @ts-ignore
    ctx => _fastCheck.default.assert(_fastCheck.default.property(...arbs, (...as) => self(as, ctx))), timeout);
  }
  const arbs = _fastCheck.default.record(Object.keys(arbitraries).reduce(function (result, key) {
    result[key] = Schema.isSchema(arbitraries[key]) ? Arbitrary.make(arbitraries[key]) : arbitraries[key];
    return result;
  }, {}));
  return V.it(name,
  // @ts-ignore
  ctx => _fastCheck.default.assert(_fastCheck.default.property(arbs, (...as) => self(as[0], ctx))), timeout);
};
/** @internal */
exports.prop = prop;
const layer = (layer_, options) => (...args) => {
  const memoMap = options?.memoMap ?? Effect.runSync(Layer.makeMemoMap);
  const scope = Effect.runSync(Scope.make());
  const runtimeEffect = Layer.toRuntimeWithMemoMap(layer_, memoMap).pipe(Scope.extend(scope), Effect.orDie, Effect.cached, Effect.runSync);
  const it = Object.assign(V.it, {
    effect: makeTester(effect => Effect.flatMap(runtimeEffect, runtime => effect.pipe(Effect.provide(runtime), Effect.provide(TestEnv)))),
    prop,
    scoped: makeTester(effect => Effect.flatMap(runtimeEffect, runtime => effect.pipe(Effect.scoped, Effect.provide(runtime), Effect.provide(TestEnv)))),
    live: makeTester(effect => Effect.flatMap(runtimeEffect, runtime => Effect.provide(effect, runtime))),
    scopedLive: makeTester(effect => Effect.flatMap(runtimeEffect, runtime => effect.pipe(Effect.scoped, Effect.provide(runtime)))),
    flakyTest,
    layer(nestedLayer, options) {
      return layer(Layer.provideMerge(nestedLayer, layer_), {
        ...options,
        memoMap
      });
    }
  });
  if (args.length === 1) {
    V.beforeAll(() => runPromise()(Effect.asVoid(runtimeEffect)), options?.timeout ? Duration.toMillis(options.timeout) : undefined);
    V.afterAll(() => runPromise()(Scope.close(scope, Exit.void)), options?.timeout ? Duration.toMillis(options.timeout) : undefined);
    return args[0](it);
  }
  return V.describe(args[0], () => {
    V.beforeAll(() => runPromise()(Effect.asVoid(runtimeEffect)), options?.timeout ? Duration.toMillis(options.timeout) : undefined);
    V.afterAll(() => runPromise()(Scope.close(scope, Exit.void)), options?.timeout ? Duration.toMillis(options.timeout) : undefined);
    return args[1](it);
  });
};
/** @internal */
exports.layer = layer;
const effect = exports.effect = /*#__PURE__*/makeTester( /*#__PURE__*/Effect.provide(TestEnv));
/** @internal */
const scoped = exports.scoped = /*#__PURE__*/makeTester( /*#__PURE__*/(0, _Function.flow)(Effect.scoped, /*#__PURE__*/Effect.provide(TestEnv)));
/** @internal */
const live = exports.live = /*#__PURE__*/makeTester(_Function.identity);
/** @internal */
const scopedLive = exports.scopedLive = /*#__PURE__*/makeTester(Effect.scoped);
/** @internal */
const flakyTest = (self, timeout = Duration.seconds(30)) => (0, _Function.pipe)(Effect.catchAllDefect(self, Effect.fail), Effect.retry((0, _Function.pipe)(Schedule.recurs(10), Schedule.compose(Schedule.elapsed), Schedule.whileOutput(Duration.lessThanOrEqualTo(timeout)))), Effect.orDie);
exports.flakyTest = flakyTest;
